### PR TITLE
Amends to allow saving and loading of compiled objects

### PR DESCRIPTION
**Pull request by tjmckinley**
Hi Stefan,

One issue I've just realised is that since the ``compile_model`` object creates a custom class, if the model object is saved out and then loaded back into a new working environment, then the custom class is not recognised and the model won't run. I'm putting forward a solution here that adds an additional slot (``comp_name``) to the ``SimInf_model`` class, which contains the name of the shared library that is created using ``compile_model``. The latter then throws out a ``SimInf_model`` object, rather than a custom class. The ``run`` method for ``SimInf_model`` objects then simply checks whether the ``comp_name`` slot is not empty and proceeds from there.

A reproducible example is this problem is given here: [test.txt](https://github.com/stewid/SimInf/files/2305847/test.txt)

With the new changes installed, this problem should go away as in here: [test_new.txt](https://github.com/stewid/SimInf/files/2305902/test_new.txt)

I've set a prototype for the new slot, so it shouldn't conflict with any of your other functions that initialise ``SimInf_model`` objects.

Cheers,

TJ

### CODE REVIEW

File: R/compile_model.R [hunk 1]
```diff
@@ -17,18 +17,17 @@
 ##' Function to compile custom \code{\link{SimInf_model}}.
```
Comment (stewid) [inline R/compile_model.R:71]: I would like to keep the `SimInf_model` class unchanged and instead define a new class (that contains a `SimInf_model`) to handle the compiled model. It's then possible to create a run function to handle the new class.   

```
##' Class to handle a compiled custom \code{\link{SimInf_model}}
##'
##' @section Slots:
##' \describe{
##'   \item{filename}{
##'     Character vector of length 1 containing name of shared library.
##'     This is used internally to deal with models that have been
##'     compiled using \code{compile_model}.
##'   }
##'   \item{filename}{
##'     A character vector of length 1 specifying the name of the
##'     custom model.
##'   }
##' }
##' @include SimInf_model.R
##' @export
setClass("SimInf_model_dll",
         contains = c("SimInf_model"),
         slots = c(filename = "character",
                   name     = "character"))

##' @rdname run
##' @export
setMethod("run",
          signature(model = "SimInf_model_dll"),
          function(model, threads, solver)
          {
              solver <- match.arg(solver)

              ## Check that SimInf_model contains all data structures
              ## required by the siminf solver and that they make sense
              validObject(model);

              dll <- dyn.load(model@filename)
              on.exit(dyn.unload(model@filename), add = TRUE)

              ## Create expression to parse
              expr <- ".Call(dll$SimInf_model_run, model, threads, solver)"

              ## Run the model. Re-throw any error without the call
              ## included in the error message to make it cleaner.
              tryCatch(eval(parse(text = expr)), error = function(e) {
                  stop(e$message, call. = FALSE)
              })
          }
)
```

File: R/compile_model.R [hunk 2]
```diff
@@ -95,11 +94,14 @@ compile_model <- function(model, name) {
...
+        model@comp_name <- filename
     } else {
         stop("No C code to compile")
     }
-    ## output new class and model
-    setClass(filename, contains = c("SimInf_model"), where = topenv(parent.frame()))
-    as(model, filename)
+    
+    ## output revised model
+    model
```
Comment (stewid) [inline R/compile_model.R]: It's now possible keep the C_code slot unchanged and to return the `SimInf_model_dll` class

```
    ## output revised model
    model <- as(model, "SimInf_model_dll")
    model@filename <- lib
    model@name  <- name
    model
```

**Review by stewid — COMMENTED** [review]
Hi TJ,

Thanks for identifying the problem. I would like to keep the SimInf_model class unchanged and instead define a new class and a run function for the new class.

File: R/compile_model.R (hunk 1 above)
Comment (tjmckinley) [reply R/compile_model.R:71]: Great idea. Thanks Stefan. Let me implement those changes and update the pull request.

**Comment by tjmckinley** [comment]
Good idea. Let me make the changes and update the request.

**Comment by tjmckinley** [comment]
Hi Stefan,

I've made the proposed amendments. This has also meant that we can roll-back the changes to the earlier ``run()`` method for ``SimInf_model`` objects. As you say, we now don't have to remove the ``C_code`` slot, which also means that we can re-compile any object at any time. This will be really useful if say, users share a ``SimInf_model_dll`` object between different operating systems, where the shared libraries use different suffixes and thus require re-compiling. Now the ``run()`` method will throw an error, and the library can simply be re-compiled through another call to ``compile_model()``.

After re-installing, see: [test_new.txt](https://github.com/stewid/SimInf/files/2307104/test_new.txt)

Cheers,

TJ

**Comment by stewid** [comment]
Many thanks @tjmckinley 

Cheers

**Comment by tjmckinley** [comment]
No worries, thanks Stefan!

**Pull request merged**